lib/jxl/dec_16bit_reconstruct.cc

```diff
@@ -0,0 +1,129 @@
+// Copyright (c) the JPEG XL Project Authors. All rights reserved.
+//
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+#include "lib/jxl/dec_16bit_reconstruct.h"
+
+#undef HWY_TARGET_INCLUDE
+#define HWY_TARGET_INCLUDE "lib/jxl/dec_16bit_reconstruct.cc"
+#include <hwy/foreach_target.h>
+#include <hwy/highway.h>
+
+#include "lib/jxl/dec_xyb-inl.h"
+
+HWY_BEFORE_NAMESPACE();
+namespace jxl {
+namespace HWY_NAMESPACE {
+void FinalizeImageRect16(Image3F* input_image, const Rect& input_rect,
+                         ImageF* alpha, const Rect& alpha_rect,
+                         PassesDecoderState* dec_state, size_t thread,
+                         const Rect& frame_rect) {
+  // This function is very NEON-specific. As such, it uses intrinsics directly.
+#if HWY_TARGET == HWY_NEON
+  // WARNING: doing fixed point arithmetic correctly is very complicated.
+  // Changes to this function should be thoroughly tested.
+  JXL_ASSERT(SameSize(input_rect, frame_rect));
+  if (alpha) JXL_ASSERT(SameSize(alpha_rect, input_rect));
+
+  const FrameDimensions& frame_dim = dec_state->shared->frame_dim;
+
+  ImageS& storage = dec_state->fixpoint_buffer[thread];
+
+  size_t x0 = frame_rect.x0();
+  size_t x1 = std::min(x0 + frame_rect.xsize(), frame_dim.xsize_upsampled);
+  size_t xs = x1 - x0;
+
+  int16_t* xyba[4] = {};
+  xyba[0] = storage.Row(9);
+  xyba[1] = storage.Row(10);
+  xyba[2] = storage.Row(11);
+  if (alpha) xyba[3] = storage.Row(12);
+
+  for (size_t y = 0; y < frame_rect.ysize() + 2 &&
+                     y + frame_rect.y0() < frame_dim.ysize_upsampled + 2;
+       y++) {
+    // Mirroring
+    size_t yfrom =
+        y + frame_rect.y0() == 0
+            ? 0
+            : (y + frame_rect.y0() == frame_dim.ysize_upsampled + 1 ? y - 2
+                                                                    : y - 1);
+
+    ConvertToFixpoint<18>(input_rect.ConstPlaneRow(*input_image, 0, yfrom) - 1,
```

Comment:
Explain me like I'm 5 why is this 18 but the the others are 15.

Reply:
I admit this is not obvious here - there's a comment about this in dec_xyb-inl.h that states that X needs extra precision, so we use X*8 instead of X. As gaborish is linear, this is OK...